### ISSUE TITLE
File name is not inline with line number when test fails

### ISSUE THREAD

**Issue by mssimo** [ThrowTheSwitch/Unity#143]
If TEST_ASSERT fails in another file than the "test file", file name is set to this test file and line number is taken from the file where TEST_ASSERT failed.

The file name in test report should be set to the file name where TEST_ASSERT failed.

**Comment by mvandervoord** [comment]
:) I've been there. This happens sometimes when either your test runner has been improperly designed or when you have created a helper file that contains asserts that do not properly pass along the updated filename. In either case, it's fixable... I just need to know a little bit more about your setup.

Are you using the test runner generator that is included with Unity? Do you have any additional files that have assertions in them other than your test file?

**Comment by mssimo** [comment]
Yes and yes :-)

**Comment by mvandervoord** [comment]
I have a strong suspicion it's with your file of extra assertions, then. I'm assuming you've built up some sort of helper file to make cleaner assertions. Like maybe you have a type like this:

typedef struct _POINT_T {
    int x;
    int y;
} POINT_T;

And then you created a helper function like:

void TEST_ASSERT_EQUAL_POINT_T(POINT_T e, POINT_T a) {
    TEST_ASSERT_EQUAL_INT(e.x, a.x);
    TEST_ASSERT_EQUAL_INT(e.y, a.y);
}

Then, in your test file, you make use of this new assertion. Unfortunately, when you do that, it would produce a problem similar to the one that you described. It would have the correct line number of the assertion, but the filename would be your original test file. Definitely NOT helpful.

It all comes from C being a somewhat limited language when things go wrong. It doesn't keep track of any type of stack trace on its own, so when we tell it we've failed, it only has the context we've given it to that point.

Fortunately, there is a solution... it just requires a little bit more work. (Fun aside: This is how CMock also handles this issue). This solution is going to make an assumption that the most important piece of information is the line number of your TEST that failed, so it approaches things from that viewpoint. If that isn't true, we can work through options in the other direction... this is usually the best approach, though.

When you create your helper functions, you need to add the ability for them to pass in the line number secretly. So we update our helper function slightly:

void TestAssertEqualPOINT_T(POINT_T e, POINT_T a, UNITY_LINE_TYPE line) {
    UNITY_TEST_ASSERT_EQUAL_INT(e.x, a.x, line, "x field");
    UNITY_TEST_ASSERT_EQUAL_INT(e.y, a.y, line, "y field");
}
#define TEST_ASSERT_EQUAL_POINT_T(e,a) TestAssertEqualPOINT_T(e,a,__LINE__)

There are a few things to notice here. First, we have created a macro which is the thing you are actually going to call. It's going to pass in the line number automatically for you using the standard C __LINE__ define (yes, that's two underscores before and after).

Second, notice we are calling standard unity macros with a UNITY_ prefix. These versions of the macros accepts a line number and a message all the time, unlike the normal calls which handle these things for you. 

Finally, we're showing off which specific thing failed in our helper by using the message fields. This makes for much nicer looking assertions and gets the information we need into the failure.

There ya go. That's how CMock tackles this problem and how I personally create helper functions. Hopefully this is useful to you. If your situation is different, feel free to say so and we can figure something out!

Mark

**Comment by mssimo** [comment]
Yes that seems to provide a workaround. But there are a few cons:

1) Need to learn a new set of macros and the test code is not consistent any more.
2) Need to add extra macros

So how come __FILE__ was not included into UNIT_TEST_ASSERT macros?

**Comment by mvandervoord** [comment]
Matti:

It's not a workaround, really. It's the behavior we built into it on
purpose. As you point out, we could have just added the FILE macro to each
assertion. This would have increased the overhead of every TEST_ASSERT,
which isn't ideal, but isn't the only reason we did it this way. The
primary reason is that this supports the ability for end-users to create
custom asserts that act exactly like Unity's asserts. Yes, it requires them
to learn some new macros (the first half of point 1 that you point out),
but this seemed worthwhile for those who wanted to customize their Unity
toolset like this.

Set up this way, developers have the ability to write assertions for their
own types and whatnot, and they can actually be automatically recognized
and used in CMock just like all the base types. This was a huge boost in
productivity for us. Up until that point, any custom type in CMock was a
straightup memory comparison. This wasn't very helpful when things went
wrong.

So, I think I've covered how we got here. But that doesn't mean we can't
improve things further and make it better for everyone. There are quite a
few Unity users who aren't also CMock users (I'm guessing yourself
included). For these people, it would be great if we could simplify the
helper-writing experience and stick with the base assertions, yes?

It can be done. I guess I'm just trying to figure out if the tradeoffs make
it worth it. It would mean that every Unity function would have to be
updated to accept an additional argument, a pointer to a const char*
containing the filename. They would each check the ptr. If it's null, they
would use the default filename (necessary for those who ARE creating custom
assertions or using CMock). If it's not null, it would use that filename
instead. CMock would also have to be updated to push an extra NULL for each
assertion, as would anyone who has written custom assertions. So it's a lot
of small changes in places and it increases our memory footprint
significantly... but it definitely would make things easier.

I think I'll leave this topic open to see if we have other people who are
interested in this functionality and/or have different solutions for fixing
it. At this point, though, I think it's worth waiting to see what kind of
feedback we get before implementing. I fear we'd be making it a worse
experience for most (slower and larger tests) while only benefiting a small
minority.

Feel free to disagree, Matti. This isn't a dictatorship. I'm definitely
open to hearing counter-points. Anyone else have thoughts?
